src/corehost/cli/fxr/fx_resolver.cpp

```diff
@@ -299,18 +333,61 @@ StatusCode fx_resolver_t::soft_roll_forward_helper(
 //      but there's no work to throw away/retry yet.
 StatusCode fx_resolver_t::soft_roll_forward(
     const fx_reference_t fx_ref,
-    bool newest_is_hard_roll_forward)
+    bool fx_is_hard_resolved)
```

Comment:
😕 Why are we calling `soft_roll_forward` when `fx_is_hard_resolved == true`?  

Doc above doesn't really explain it clearly.  I suppose if it is `fx_is_hard_resolved == false`, it will trigger a retry, but if it is `fx_is_hard_resolved == true` it will trigger an error.

I wonder if this can be simplified by eliniminating internal state changes and returning the a new `fx_reference_t`

i.e. 
```C++
fx_reference_t fx_resolver_t::soft_roll_forward(const fx_reference_t fx_ref) const;
```

Reply:
Maybe we should rename "soft roll forward" to "reconcile framework reference" or something similar. The function takes the passed in framework reference and reconciles it with the current effective framework reference (`m_newest_references` in the code).

The reconciling means:
* Check that it's compatible (`is_roll_forward_compatible`)
  * If it's not fail
* Reconcile the references and produce the new effective reference
  * Take the higher version
  * Reconcile the roll forward settings (take the more strict)
* Update the current effective framework reference to the new reconciled one (update `m_newest_references`)

The `fx_is_hard_resolved` means basically:
"The current effective reference for the framework in question has already been used to resolve the framework on disk".
So if the function ends up updating the current effective framework reference, it needs to trigger the retry, since the "resolve framework on disk" we've done already is no longer valid.

Possible improvements:
* Rename stuff to make more sense... I guess I was basically waiting for the review to see what makes sense and what is confusing. I'll think about better names and probably do some renaming.
* As you suggest, make the function return the new effective reference but without any side effects. Then the caller would have to check that if it's different and we've already resolved framework on disk, then trigger the rety.

Reply:
https://github.com/dotnet/core-setup/pull/5691/commits/464a397b0878ebc2a62e07dfeca02300d42f4306 addresses this.

Made the functions pure (no side effects).
Got rid of the weird "soft-roll-forward" and "hard-roll-forward" terms.
Some other renames/refactorings - the code should be much easier to follow now.